pkg/logger/ecs_error_log.go

```diff
@@ -0,0 +1,188 @@
+	}
+	l.size = 0
+
+	errorNum := len(l.queue)
+	// iterate errors in queue, if the old error's message same as the new error's message, skip it,
+	// otherwise write it into the file and queue
+	for i := 0; i < errorNum; i++ {
+		e := l.queue[0]
+		if e.Message != newError.Message {
+			n, err := l.errorFile.Write([]byte(fmt.Sprintf("{%+v, Level:%+v, Caller:%+v, Message:%+v}\r\n",
+				e.Time, e.Level, e.Caller, e.Message)))
+			if err != nil {
+				log.Printf("could not write error into error log file, err: %v", err)
+			}
+			l.size += n
+			l.queue = append(l.queue, e)
+		}
+		l.queue = l.queue[1:]
+	}
+	// delete the old errors if old errors plus new error exceed the max size
+	if l.size+len(newErrorLog) > l.ErrorFileMaxSize {
+		l.rotate(newErrorLog)
+	}
+
+	l.queue = append(l.queue, newError)
+	n, err := l.errorFile.Write(newErrorLog)
+	if err != nil {
+		log.Printf("could not write new error into error log file, err: %v", err)
+	}
+	l.size += n
+}
+
+// rotate func could delete the old logs to make sure error log file less than 1 KB
+func (l *ECSErrorLogger) rotate(newErrorLog []byte) {
+	file, err := os.OpenFile(l.ErrorFilePath, os.O_CREATE|os.O_WRONLY|os.O_TRUNC|os.O_APPEND, 0777)
+	if err != nil {
+		log.Printf("could not open error log file during rotation, err: %v", err)
+	}
+
+	err = file.Truncate(0)
+	if err != nil {
+		log.Printf("could not truncate error log file during rotation, err: %v", err)
+	}
+
+	errorNum := len(l.queue)
+	for i := 0; i < errorNum; i++ {
+		currentError := l.queue[0]
+		currentErrorLog := []byte(fmt.Sprintf("{%+v, Level:%+v, Caller:%+v, Message:%+v}\r\n",
+			currentError.Time, currentError.Level, currentError.Caller, currentError.Message))
+		if l.size+len(newErrorLog) > l.ErrorFileMaxSize {
+			l.size -= len(currentErrorLog)
+		} else {
+			l.queue = append(l.queue, currentError)
+			_, err = file.Write(currentErrorLog)
+			if err != nil {
+				log.Printf("could not write error into error log file during rotation, err: %v", err)
+			}
+		}
+		l.queue = l.queue[1:]
+	}
+}
+
+// processTimeout function could delete the expired error log from the file based on the ErrorLogMaxAge
+func (l *ECSErrorLogger) processTimeout() {
+	currentTime := time.Now()
+	file, err := os.OpenFile(l.ErrorFilePath, os.O_CREATE|os.O_WRONLY|os.O_TRUNC|os.O_APPEND, 0777)
+	if err != nil {
+		log.Printf("could not open error log file during process timeout, err: %v", err)
+	}
+
+	err = file.Truncate(0)
+	if err != nil {
+		log.Printf("could not truncate error log file during rotation, err: %v", err)
+	}
+	l.size = 0
+
+	errorNum := len(l.queue)
+	for i := 0; i < errorNum; i++ {
+		e := l.queue[0]
+		if e.Time.Add(time.Duration(l.ErrorLogMaxAge) * time.Hour).After(currentTime) {
+			n, err := file.Write([]byte(fmt.Sprintf("{%+v, Level:%+v, Caller:%+v, Message:%+v}\r\n",
+				e.Time, e.Level, e.Caller, e.Message)))
+			if err != nil {
+				log.Printf("could not write error into error log file during process timeout, err: %v", err)
+			}
```

Comment:
One thing I am noticing in this code. There are 3 functions which write to the file. So the code for writing is duplicated 3 times. That is not ideal from an organization and maintainability standpoint. 

It would be ideal if each of these functions only manipulates the queue of messages. And then there is one separate function for writing all contents of the queue to the file. And then each function can call that function to write.

Reply:
but you can see that for those 3 functions, though they all write to the file, it has some different logic during the iteration of the queue

Reply:
> but you can see that for those 3 functions, though they all write to the file, it has some different logic during the iteration of the queue

Yeah and that's exactly my point. The custom logic for iterating over the queue and manipulating it should be in each function. Each function does its job to manipulate the queue. It gets the queue in the correct final state. And then you call a common utility function to write the contents of the queue to the file.